pkg/telemetry/span.go

```diff
@@ -0,0 +1,556 @@
+		TraceFlags: psc.TraceFlags(),
+	}
+
+	s := &Span{
+		parent:   psc,
+		name:     so.SpanName(),
+		scope:    instrumentation.Scope{Name: so.SpanScope()},
+		resource: so.Resource(),
+		start:    so.Timestamp(),
+		attrs:    so.Attributes(),
+		events:   []tracesdk.Event{},
+		links:    so.Links(),
+		status:   tracesdk.Status{Code: codes.Unset},
+		conf:     sconf,
+		kind:     so.SpanKind(),
+		dedup:    so.Dedup(),
+	}
+
+	return trace.ContextWithSpan(ctx, s), s
+}
+
+// span is an attempt to mimic the otel span data structure following the protobuf spec at
+// https://github.com/open-telemetry/opentelemetry-proto/blob/v1.1.0/opentelemetry/proto/trace/v1/trace.proto
+//
+// Due to the limitations of the otel lib's API interface, we can't reconstruct spans over boundaries,
+// and in order to make sure the execution data looks like how it looks from the SDK side,
+// we'll need to work around the otel library and have slightly different way of working with the data
+//
+// This file is an attempt to make it as close as possible to official libs so we can minimize deviations.
+//
+// NOTE: to make sure it doesn't conflict the the ReadOnlySpan interface functions,
+// certain fields are named in a little weird way.
+type Span struct {
+	tracesdk.ReadWriteSpan // embeds both span interfaces
+	sync.Mutex
+
+	start time.Time
+
+	name   string
+	attrs  []attribute.KeyValue
+	status tracesdk.Status
+	events []tracesdk.Event
+	links  []tracesdk.Link
+	kind   trace.SpanKind
+
+	scope    instrumentation.Scope
+	resource *resource.Resource
+	parent   trace.SpanContext
+	conf     trace.SpanContextConfig
+
+	// dedup marks the span as a potential duplicate, which in turn
+	// can be used as an indicator to allow the span to be sent of not.
+	dedup bool
+	// Mark the span as cancelled, so it doesn't get sent out when it ends
+	cancel            bool
+	childSpanCount    int
+	droppedAttributes int
+}
+
+// Send is just an alias for End
+func (s *Span) Send() {
+	s.End()
+}
+
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+//
+//	FOOTGUN ALERT
+//
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// Cancel will mark the span as cancelled and it will not be sent out when it ends.
+// This will reset the context so if there are spans that will be created after this,
+// it doesn't create a dangling pointer.
+func (s *Span) Cancel(ctx context.Context) context.Context {
+	s.cancel = true
+	// revert the current span context back to the parent's
+	return trace.ContextWithSpanContext(ctx, s.Parent())
+}
+
+//
+// trace.ReadOnlySpan interface functions
+//
+
+func (s *Span) Name() string {
+	return s.name
+}
+
+func (s *Span) SpanContext() trace.SpanContext {
+	return trace.NewSpanContext(s.conf)
+}
+
+func (s *Span) Parent() trace.SpanContext {
+	return s.parent
+}
+
+func (s *Span) SpanKind() trace.SpanKind {
+	return s.kind
+}
+
+func (s *Span) StartTime() time.Time {
+	return s.start
+}
+
+func (s *Span) EndTime() time.Time {
+	return time.Now()
+}
```

Comment:
Can't find any references to this function in this PR, so I'm guessing this is called internally by the exporter and is part of our matching the otel interface.

If so, is this correct? I'm wondering if `time.Now()` will mean that the end time of the span is set to the moment it is exported rather than when we ended it with `.End()` or `.Send()`. For batched spans especially, this could cause some weird timings.

Reply:
addressed at [49ec2da](https://github.com/inngest/inngest/pull/1248/commits/49ec2daf7873519944b10cdeb50d5e1c7ffe5340)